aggregator/apps/api/v1/api_client.py

```diff
@@ -92,11 +107,86 @@ def get_data_for_address(self, slug):
         else:
             wcivf_url = f'{settings.WCIVF_BASE_URL}candidates_for_postcode/?postcode={wdiv_json["addresses"][0]["postcode"]}'
 
-        wcivf_response = asyncio.ensure_future(fetch(wcivf_url))
-        self.loop.run_until_complete(wcivf_response)
-        wcivf_result = wcivf_response.result()
-        if wcivf_result["status"] >= 400:
-            raise ApiError(wcivf_result["status"], wcivf_result["json"]["detail"])
+        wcivf_result = proxy_single_request(self.loop, wcivf_url, self.wcivf_params)
         wcivf_json = wcivf_result["json"]
 
         return wdiv_json, wcivf_json
+
+
+class EEApiClient(AsyncApiClient):
+    def clean_query_params(self, params):
+        """
+        Take a django QueryDict object and return another QueryDict
+        containing only the specified keys.
+
+        We should use this both:
+        - before passing a query string from devs.DC to EE and
+        - before passing a URL with a query string from EE
+          back to the devs.DC user.
+
+        There are several things this acheives:
+        1. There are some params that the devs.DC API accepts
+           that we don't want to pass through to the EE API (e.g: auth_token)
+        2. There are some params supported by the EE API
+           that we don't want to expose to dev.DC users (e.g: postcode)
+        3. In future, if we wanted to pass additional params to EE
+           (e.g: a tracking param or API key, like we do with WDIV),
+           they'll come back reflected in the next/previous URLs.
+           Cleaning prevents us exposing those params to the devs.DC user.
```

Comment:
I think I've explained this OK here, but just to clarify what this third point with an example:

With WDIV and WCIVF, we pass `utm_medium=devs.DC API` in the querystring for stats purposes because WDIV and WCIVF store that in the `LoggedPostcode` table. I haven't bothered with EE because EE doesn't log it, but if we did that, the `next` URL in the pagination links will come back as `https://elections.democracyclub.org.uk/api/elections/?limit=100&offset=100&utm_medium=devs.DC API`. Same would go for an API key, or whatever. This allows us to safely modify the querystring at this layer before we forward it on to EE without it impacting on the user. We'd just expose the pagination link as `https://developers.democracyclub.org.uk/api/v1/elections/?limit=100&offset=100`

Reply:
👍